api/app/Models/Article.php

```diff
@@ -85,6 +87,25 @@ protected static function boot()
         });
     }
 
+    /**
+     * @param string $id article_id or permalink
+     * @return Article
+     */
+    public function findByPermalink($id)
+    {
+        //if the id is a uuid, try that or fail.
+        if (Uuid::isValid($id)) {
+            return parent::findOrFail($id);
+        }
+
+        //otherwise attempt treat the id as a permalink and first try the model, then try the history
+        try {
+            return $this->where('permalink', '=', $id)->firstOrFail();
+        } catch (ModelNotFoundException $e) { //id or permalink not found, try permalink history
+            return ArticlePermalink::findOrFail($id)->article;
+        }
+    }
+
```

Comment:
then we will have to duplicate that for ArticleMetaController and ArticlePermalinkContoller